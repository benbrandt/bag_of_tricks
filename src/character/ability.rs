use crate::dice_roller::{roll_dice, Die};
use rand::Rng;

/// Value of a base ability score.
#[derive(Debug)]
pub(crate) struct AbilityScore(pub(crate) i32);

impl AbilityScore {
    /// Generate a new ability score based on dice rolls
    fn new(rng: &mut impl Rng) -> Self {
        // Roll 4 d6's
        let mut rolls = roll_dice(rng, Die::D6, 4);
        // Reverse sort, highest to lowest
        rolls.sort_by(|a, b| b.roll.cmp(&a.roll));
        // Sum top 3
        let score = rolls.drain(0..3).fold(0, |acc, d| acc + d.roll);
        Self(score)
    }

    /// Return modifier based on ability score.
    const fn modifier(&self) -> i32 {
        // Lower value to closest even number, subtract by 10 and divide by two
        (self.0 - self.0 % 2 - 10) / 2
    }
}

/// Full set of ability scores a character could have
#[derive(Debug)]
pub(crate) struct AbilityScores {
    pub(crate) charisma: AbilityScore,
    pub(crate) constitution: AbilityScore,
    pub(crate) dexterity: AbilityScore,
    pub(crate) intelligence: AbilityScore,
    pub(crate) strength: AbilityScore,
    pub(crate) wisdom: AbilityScore,
}

impl AbilityScores {
    /// Generate a set of ability scores for a character
    pub(crate) fn new(rng: &mut impl Rng) -> Self {
        Self {
            charisma: AbilityScore::new(rng),
            constitution: AbilityScore::new(rng),
            dexterity: AbilityScore::new(rng),
            intelligence: AbilityScore::new(rng),
            strength: AbilityScore::new(rng),
            wisdom: AbilityScore::new(rng),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_ability_score_new() {
        let mut rng = Pcg64::from_entropy();
        let score = AbilityScore::new(&mut rng);
        assert!(score.0 >= 3 && score.0 <= 18);
    }

    #[test]
    fn test_ability_score_avg() {
        let mut rng = Pcg64::from_entropy();
        let average = (0..100).fold(0, |acc, _| acc + AbilityScore::new(&mut rng).0) as f64 / 100.0;
        // Comparison based on http://rumkin.com/reference/dnd/diestats.php
        assert!(12.24 - 2.847 < average && average < 12.24 + 2.847);
    }

    #[test]
    fn test_ability_score_modifier() {
        assert_eq!(AbilityScore(1).modifier(), -5);
        assert_eq!(AbilityScore(2).modifier(), -4);
        assert_eq!(AbilityScore(3).modifier(), -4);
        assert_eq!(AbilityScore(4).modifier(), -3);
        assert_eq!(AbilityScore(5).modifier(), -3);
        assert_eq!(AbilityScore(6).modifier(), -2);
        assert_eq!(AbilityScore(7).modifier(), -2);
        assert_eq!(AbilityScore(8).modifier(), -1);
        assert_eq!(AbilityScore(9).modifier(), -1);
        assert_eq!(AbilityScore(10).modifier(), 0);
        assert_eq!(AbilityScore(11).modifier(), 0);
        assert_eq!(AbilityScore(12).modifier(), 1);
        assert_eq!(AbilityScore(13).modifier(), 1);
        assert_eq!(AbilityScore(14).modifier(), 2);
        assert_eq!(AbilityScore(15).modifier(), 2);
        assert_eq!(AbilityScore(16).modifier(), 3);
        assert_eq!(AbilityScore(17).modifier(), 3);
        assert_eq!(AbilityScore(18).modifier(), 4);
        assert_eq!(AbilityScore(19).modifier(), 4);
        assert_eq!(AbilityScore(20).modifier(), 5);
    }

    #[test]
    fn test_ability_scores() {
        let mut rng = Pcg64::from_entropy();
        let AbilityScores {
            charisma,
            constitution,
            dexterity,
            intelligence,
            strength,
            wisdom,
        } = AbilityScores::new(&mut rng);
        assert!(charisma.0 >= 3 && charisma.0 <= 18);
        assert!(constitution.0 >= 3 && constitution.0 <= 18);
        assert!(dexterity.0 >= 3 && dexterity.0 <= 18);
        assert!(intelligence.0 >= 3 && intelligence.0 <= 18);
        assert!(strength.0 >= 3 && strength.0 <= 18);
        assert!(wisdom.0 >= 3 && wisdom.0 <= 18);
    }
}
