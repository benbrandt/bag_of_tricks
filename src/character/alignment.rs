use std::fmt;

use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Attitude {
    Chaotic,
    Lawful,
    Neutral,
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Morality {
    Evil,
    Good,
    Neutral,
}

#[derive(Deserialize, Serialize)]
pub(crate) struct Alignment(Attitude, Morality);

impl Alignment {
    pub(crate) fn gen(rng: &mut impl Rng) -> Self {
        let attitude = Attitude::iter().choose(rng).unwrap();
        let morality = Morality::iter().choose(rng).unwrap();
        Self(attitude, morality)
    }
}

impl fmt::Display for Alignment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self(Attitude::Neutral, Morality::Neutral) => write!(f, "Neutral"),
            Self(a, m) => write!(f, "{} {}", a, m),
        }
    }
}
