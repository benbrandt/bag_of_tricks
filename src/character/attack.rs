use std::fmt;

use strum_macros::Display;

use crate::dice_roller::RollCmd;

#[allow(dead_code)]
#[derive(Display)]
pub(crate) enum DamageType {
    Acid,
    Bludgeoning,
    Cold,
    Fire,
    Force,
    Lightning,
    Necrotic,
    Piercing,
    Poison,
    Psychic,
    Radiant,
    Slashing,
    Thunder,
}

pub(crate) struct Damage {
    pub(crate) damage_type: DamageType,
    pub(crate) modifier: i16,
    pub(crate) roll: RollCmd,
}

impl fmt::Display for Damage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}{:+} {}", self.roll, self.modifier, self.damage_type)
    }
}

// pub(crate) enum Hit {
//     Roll(u8),
//     DC(AbilityScoreType, i16),
// }

// impl fmt::Display for Hit {
//     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
//         match self {
//             Self::Roll(m) => write!(f, "{:+}", m),
//             Self::DC(a, s) => write!(f, "{} {}", s, a),
//         }
//     }
// }

// pub(crate) struct Attack<'a> {
//     pub(crate) citation: Citation,
//     pub(crate) hit: Hit,
//     pub(crate) damage: Damage,
//     pub(crate) name: &'a str,
//     pub(crate) range: &'a str,
// }

// impl fmt::Display for Attack<'_> {
//     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
//         writeln!(
//             f,
//             "{:20} {:20} {:20} {:20} {}",
//             self.name, self.range, self.hit, self.damage, self.citation
//         )
//     }
// }
