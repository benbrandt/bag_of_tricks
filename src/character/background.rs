mod acolyte;
mod charlatan;
mod criminal;
mod entertainer;
mod folk_hero;
mod guild_artisan;
mod hermit;
mod noble;
mod outlander;
mod sage;
mod sailor;
mod soldier;
mod urchin;

use std::fmt;

use entertainer::Entertainer;
use folk_hero::FolkHero;
use rand::{
    prelude::{Distribution, SliceRandom},
    Rng,
};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use crate::citation::Citations;

use self::{
    acolyte::Acolyte, charlatan::Charlatan, criminal::Criminal, guild_artisan::GuildArtisan,
    hermit::Hermit, noble::Noble, outlander::Outlander, sage::Sage, sailor::Sailor,
    soldier::Soldier, urchin::Urchin,
};

use super::{
    ability::{weighted_modifiers_dist, AbilityScores, Skill},
    alignment::{AlignmentInfluences, Attitude, Morality},
    backstory::Backstory,
    equipment::StartingEquipment,
    features::Features,
    languages::Languages,
    proficiencies::Proficiencies,
};

/// Types of alignment influence from personality traits
#[derive(Clone, Copy, Deserialize, Display, Serialize)]
pub(crate) enum Influence {
    Any,
    Chaotic,
    Evil,
    Good,
    Lawful,
    Neutral,
}

impl AlignmentInfluences for Influence {
    /// Return attitude influence from personality.
    /// Doubled because personality should be a major indicator of alignment.
    fn attitude(&self) -> Vec<Attitude> {
        match self {
            Self::Chaotic => vec![Attitude::Chaotic; 2],
            Self::Lawful => vec![Attitude::Lawful; 2],
            Self::Neutral => vec![Attitude::Neutral; 2],
            _ => vec![],
        }
    }

    /// Return morality influence from personality.
    /// Doubled because personality should be a major indicator of alignment.
    fn morality(&self) -> Vec<Morality> {
        match self {
            Self::Good => vec![Morality::Good; 2],
            Self::Evil => vec![Morality::Evil; 2],
            Self::Neutral => vec![Morality::Neutral; 2],
            _ => vec![],
        }
    }
}

/// Description of a character's personality
#[derive(Deserialize, Serialize)]
pub(crate) struct Personality {
    bond: String,
    flaw: String,
    ideal: (String, Influence),
    traits: Vec<String>,
}

/// Return attitude/morality influences from character's ideal
impl AlignmentInfluences for Personality {
    fn attitude(&self) -> Vec<Attitude> {
        self.ideal.1.attitude()
    }

    fn morality(&self) -> Vec<Morality> {
        self.ideal.1.morality()
    }
}

impl fmt::Display for Personality {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "PERSONALITY TRAITS:")?;
        for t in &self.traits {
            writeln!(f, "{}", t)?;
        }
        writeln!(f, "IDEAL: {} ({})", self.ideal.0, self.ideal.1)?;
        writeln!(f, "BOND: {}", self.bond)?;
        writeln!(f, "FLAW: {}", self.flaw)
    }
}

/// Trait to store associated constants for each background's personality tables.
pub(crate) trait PersonalityOptions {
    /// List of 6 bonds to choose from
    const BONDS: [&'static str; 6];
    /// List of 6 flaws to choose from
    const FLAWS: [&'static str; 6];
    /// List of 6 ideals to choose from, and their corresponding alignment influence
    const IDEALS: [(&'static str, Influence); 6];
    /// List of 8 traits to choose from
    const TRAITS: [&'static str; 8];

    /// Generate personality descriptions from the associated constants
    fn gen_personality(rng: &mut impl Rng) -> Personality {
        let ideal = *Self::IDEALS.choose(rng).unwrap();
        Personality {
            bond: String::from(*Self::BONDS.choose(rng).unwrap()),
            flaw: String::from(*Self::FLAWS.choose(rng).unwrap()),
            ideal: (String::from(ideal.0), ideal.1),
            traits: Self::TRAITS
                .choose_multiple(rng, 2)
                .map(|s| String::from(*s))
                .collect(),
        }
    }
}

/// Trait for backgrounds to build from
#[typetag::serde(tag = "type")]
pub(crate) trait Background:
    Backstory + Citations + Features + Languages + Proficiencies + StartingEquipment + fmt::Display
{
    /// Generate a new instance of the background
    fn gen(rng: &mut impl Rng) -> (Box<dyn Background>, Personality)
    where
        Self: Sized;

    /// Return list of skills background gives proficiency in
    fn skills() -> Vec<Skill>
    where
        Self: Sized;
}

/// List of currently supported background options
#[derive(EnumIter)]
pub(crate) enum BackgroundOption {
    Acolyte,
    Charlatan,
    Criminal,
    Entertainer,
    FolkHero,
    GuildArtisan,
    Hermit,
    Noble,
    Outlander,
    Sage,
    Sailor,
    Soldier,
    Urchin,
}

impl BackgroundOption {
    /// Choose a random background option, weighted by proficiency bonuses, and map to corresponding generator
    pub(crate) fn gen(
        rng: &mut impl Rng,
        abilities: &AbilityScores,
    ) -> (Box<dyn Background>, Personality) {
        let options: Vec<BackgroundOption> = Self::iter().collect();
        let modifiers = Self::iter().map(|o| o.skill_modifier(abilities));
        let dist = weighted_modifiers_dist(modifiers);
        match options.get(dist.sample(rng)).unwrap() {
            Self::Acolyte => Acolyte::gen(rng),
            Self::Charlatan => Charlatan::gen(rng),
            Self::Criminal => Criminal::gen(rng),
            Self::Entertainer => Entertainer::gen(rng),
            Self::FolkHero => FolkHero::gen(rng),
            Self::GuildArtisan => GuildArtisan::gen(rng),
            Self::Hermit => Hermit::gen(rng),
            Self::Noble => Noble::gen(rng),
            Self::Outlander => Outlander::gen(rng),
            Self::Sage => Sage::gen(rng),
            Self::Sailor => Sailor::gen(rng),
            Self::Soldier => Soldier::gen(rng),
            Self::Urchin => Urchin::gen(rng),
        }
    }

    /// Max skill modifier of background for weighting
    fn skill_modifier(&self, abilities: &AbilityScores) -> i16 {
        let skills = match self {
            Self::Acolyte => Acolyte::skills(),
            Self::Charlatan => Charlatan::skills(),
            Self::Criminal => Criminal::skills(),
            Self::Entertainer => Entertainer::skills(),
            Self::FolkHero => FolkHero::skills(),
            Self::GuildArtisan => GuildArtisan::skills(),
            Self::Hermit => Hermit::skills(),
            Self::Noble => Noble::skills(),
            Self::Outlander => Outlander::skills(),
            Self::Sage => Sage::skills(),
            Self::Sailor => Sailor::skills(),
            Self::Soldier => Soldier::skills(),
            Self::Urchin => Urchin::skills(),
        };
        skills
            .iter()
            .map(|s| abilities.modifier(s.ability_score_type()))
            .max()
            .unwrap_or(0)
    }
}
