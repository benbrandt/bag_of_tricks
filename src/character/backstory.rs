pub(crate) trait Backstory {
    fn backstory(&self) -> Vec<String> {
        vec![]
    }
}
