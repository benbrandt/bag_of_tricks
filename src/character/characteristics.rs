use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use std::{fmt, ops::RangeInclusive};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use crate::dice_roller::RollCmd;

/// An range of ages a given adventurer could be.
pub(crate) struct AgeRange(pub(crate) RangeInclusive<u16>);

impl AgeRange {
    /// Generate a random age within the given range.
    fn gen(&self, rng: &mut impl Rng) -> u16 {
        rng.gen_range(self.0.clone())
    }
}

/// Really only here to help decide on names, not core to a character choice.
#[derive(Deserialize, Display, EnumIter, Serialize)]
pub(crate) enum Gender {
    Female,
    Male,
}

impl Gender {
    /// Choose a random gender
    fn gen(rng: &mut impl Rng) -> Self {
        Gender::iter().choose(rng).unwrap()
    }
}

/// Weight modifier
#[derive(Debug, PartialEq)]
pub(crate) enum WeightMod {
    /// Fixed additional weight
    Fixed(u16),
    /// Role to find additional weight
    Roll(RollCmd),
}

/// Convert feet + inches to just inches (allow for better readability in tables)
pub(crate) const fn in_inches(feet: u8, inches: u8) -> u8 {
    feet * 12 + inches
}

/// Data from a table to generate a height and weight for a given character.
#[derive(Debug, PartialEq)]
pub(crate) struct HeightAndWeightTable {
    /// Base height for this character
    pub(crate) base_height: u8,
    /// Base weight for this character
    pub(crate) base_weight: u16,
    /// Modifier to apply to the height
    pub(crate) height_mod: RollCmd,
    /// Modifier to apply to the weight (informed by height)
    pub(crate) weight_mod: WeightMod,
}

impl HeightAndWeightTable {
    /// Generate a random height and weight for a character
    pub(crate) fn gen(&self, rng: &mut impl Rng) -> (usize, usize) {
        let h = self.height_mod.roll(rng).total();
        // Weight modifier is multiplied by height
        let w = h * match &self.weight_mod {
            WeightMod::Fixed(f) => usize::from(*f),
            WeightMod::Roll(r) => r.roll(rng).total(),
        };
        (
            usize::from(self.base_height) + h,
            usize::from(self.base_weight) + w,
        )
    }
}

/// Size of character (there are more options for monsters)
#[derive(Deserialize, Display, EnumIter, Serialize)]
pub(crate) enum Size {
    Small,
    Medium,
}

/// Types of movement speeds
#[derive(Clone, Copy, Debug, Deserialize, EnumIter, PartialEq, Serialize)]
pub(crate) enum Speed {
    Climbing(u8),
    Swimming(u8),
    Walking(u8),
}

impl fmt::Display for Speed {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Climbing(s) => write!(f, "Climbing Speed: {}ft", s),
            Self::Swimming(s) => write!(f, "Swimming Speed: {}ft", s),
            Self::Walking(s) => write!(f, "Walking Speed: {}ft", s),
        }
    }
}

/// Physical characteristics about a character.
#[derive(Deserialize, Serialize)]
pub(crate) struct CharacteristicDetails {
    /// Age of the character
    pub(crate) age: u16,
    /// Base speed of the character
    pub(crate) base_speeds: Vec<Speed>,
    /// Gender of the character (only used for name choices)
    pub(crate) gender: Gender,
    /// Height of the character
    pub(crate) height: usize,
    /// Size of the character
    pub(crate) size: Size,
    /// Weight of the character
    pub(crate) weight: usize,
}

impl fmt::Display for CharacteristicDetails {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "Age: {}", self.age)?;
        writeln!(f, "Gender: {}", self.gender)?;
        writeln!(f, "Size: {}", self.size)?;
        writeln!(f, "Height: {}'{}\"", self.height / 12, self.height % 12)?;
        writeln!(f, "Weight: {} lb.", self.weight)
    }
}

/// Trait for object (usually race) to generate appropriate characteristics about a character
///
/// Separate from race since it uses associated constants (which can't be on a trait object)
pub(crate) trait Characteristics {
    /// Age range to choose from
    const AGE_RANGE: AgeRange;
    /// Character size
    const SIZE: Size;

    /// Base speeds
    fn get_base_speeds(&self) -> Vec<Speed>;

    /// Calculate the height and weight table to generate from
    fn get_height_and_weight_table(&self) -> &HeightAndWeightTable;

    /// Generate characteristics for this race
    fn gen_characteristics(&self, rng: &mut impl Rng) -> CharacteristicDetails {
        let (height, weight) = self.get_height_and_weight_table().gen(rng);
        CharacteristicDetails {
            age: Self::AGE_RANGE.gen(rng),
            base_speeds: self.get_base_speeds(),
            gender: Gender::gen(rng),
            height,
            size: Self::SIZE,
            weight,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_age_range() {
        let mut rng = Pcg64::from_entropy();
        let age = AgeRange(1..=100).gen(&mut rng);
        assert!((1..=100).contains(&age));
    }

    /// Verify that our snapshot remains the same.
    #[test]
    fn test_characteristics() {
        let mut rng = Pcg64::seed_from_u64(1);
        let characteristics = CharacteristicDetails {
            age: 100,
            base_speeds: vec![Speed::Walking(30)],
            gender: Gender::gen(&mut rng),
            height: 75,
            size: Size::Medium,
            weight: 300,
        };
        insta::assert_yaml_snapshot!(characteristics);
    }

    /// Verify that our snapshot remains the same.
    #[test]
    fn test_characteristics_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let characteristics = CharacteristicDetails {
            age: 100,
            base_speeds: vec![Speed::Walking(30)],
            gender: Gender::gen(&mut rng),
            height: 75,
            size: Size::Medium,
            weight: 300,
        };
        insta::assert_snapshot!(format!("{}", characteristics));
    }
}
