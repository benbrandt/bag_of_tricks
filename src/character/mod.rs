mod ability;

use ability::{AbilityScoreIncreases, AbilityScores};
use rand::Rng;
use std::fmt;

/// Character information
#[derive(Debug)]
pub struct Character {
    abilities: AbilityScores,
}

impl Character {
    /// Generate a new random character
    pub fn new(rng: &mut impl Rng) -> Self {
        Self {
            abilities: AbilityScores::new(rng, AbilityScoreIncreases::default()),
        }
    }
}

impl fmt::Display for Character {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.abilities)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_character_new() {
        let mut rng = Pcg64::from_entropy();
        let Character {
            abilities:
                AbilityScores {
                    charisma,
                    constitution,
                    dexterity,
                    intelligence,
                    strength,
                    wisdom,
                },
        } = Character::new(&mut rng);
        assert!(charisma.base >= 3 && charisma.base <= 18);
        assert!(constitution.base >= 3 && constitution.base <= 18);
        assert!(dexterity.base >= 3 && dexterity.base <= 18);
        assert!(intelligence.base >= 3 && intelligence.base <= 18);
        assert!(strength.base >= 3 && strength.base <= 18);
        assert!(wisdom.base >= 3 && wisdom.base <= 18);
    }
}
