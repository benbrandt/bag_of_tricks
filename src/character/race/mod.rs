pub(crate) mod dragonborn;
pub(crate) mod dwarf;

use std::fmt;

use rand::Rng;
use rand_derive::Rand;

use super::ability::AbilityScoreIncreases;

/// Shared race traits
pub(crate) trait Race: fmt::Display {
    /// Method to generate a new instance of the struct
    fn new(rng: &mut impl Rng) -> Self
    where
        Self: Sized;

    /// Returns ability score increases for the race
    fn increases(&self) -> AbilityScoreIncreases;
}

#[derive(Rand)]
enum RaceOptions {
    Dragonborn,
    Dwarf,
}

pub(crate) fn gen_race_option(rng: &mut impl Rng) -> Box<dyn Race> {
    match rng.gen::<RaceOptions>() {
        RaceOptions::Dragonborn => Box::new(dragonborn::Dragonborn::new(rng)),
        RaceOptions::Dwarf => Box::new(dwarf::Dwarf::new(rng)),
    }
}
