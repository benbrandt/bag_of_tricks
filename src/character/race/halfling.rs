use rand::prelude::IteratorRandom;
use rand::Rng;
use serde::{Deserialize, Serialize};
use std::fmt;
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use super::Race;
use crate::{
    character::{
        ability::{AbilityScore, AbilityScoreType, AbilityScores},
        characteristics::{AgeRange, Characteristics},
        features::Feature,
    },
    citation::{Book, Citation, Citations},
};

const AGE_RANGE: AgeRange = AgeRange(1..=150);
mod names {
    use rand::{prelude::IteratorRandom, Rng};

    use crate::character::characteristics::{Characteristics, Gender};

    const FAMILY: &[&str] = &[
        "Brushgather",
        "Goodbarrel",
        "Greenbottle",
        "High-hill",
        "Hilltopple",
        "Leagallow",
        "Tealeaf",
        "Thorngage",
        "Tosscobble",
        "Underbough",
    ];

    const FEMALE: &[&str] = &[
        "Andry",
        "Bree",
        "Callie",
        "Cora",
        "Euphemia",
        "Jillian",
        "Kithri",
        "Lavinia",
        "Lidda",
        "Merla",
        "Nedda",
        "Paela",
        "Portia",
        "Seraphina",
        "Shaena",
        "Trym",
        "Vani",
        "Verna",
    ];

    const MALE: &[&str] = &[
        "Alton", "Ander", "Cade", "Corrin", "Eldon", "Errich", "Finnan", "Garret", "Lindal",
        "Lyle", "Merric", "Milo", "Osborn", "Perrin", "Reed", "Roscoe", "Wellby",
    ];

    pub(crate) fn gen_name(
        rng: &mut impl Rng,
        Characteristics { gender, .. }: &Characteristics,
    ) -> String {
        let first_names = match gender {
            Gender::Female => FEMALE,
            Gender::Male => MALE,
        };
        format!(
            "{} {}",
            first_names.iter().choose(rng).unwrap(),
            FAMILY.iter().choose(rng).unwrap(),
        )
    }
}

#[derive(Debug, Deserialize, Display, EnumIter, PartialEq, Serialize)]
enum HalflingSubrace {
    Lightfoot,
    Stout,
}

#[derive(Deserialize, Serialize)]
pub(crate) struct Halfling {
    subrace: HalflingSubrace,
}

#[typetag::serde]
impl Race for Halfling {
    fn gen(rng: &mut impl Rng) -> (Box<dyn Race>, String, Characteristics) {
        let race = Box::new(Self {
            subrace: HalflingSubrace::iter().choose(rng).unwrap(),
        });
        let characteristics = Characteristics::gen(rng, &AGE_RANGE);
        let name = names::gen_name(rng, &characteristics);
        (race, name, characteristics)
    }

    fn abilities(&self) -> AbilityScores {
        AbilityScores(vec![
            AbilityScore(AbilityScoreType::Dexterity, 2),
            match self.subrace {
                HalflingSubrace::Lightfoot => AbilityScore(AbilityScoreType::Charisma, 1),
                HalflingSubrace::Stout => AbilityScore(AbilityScoreType::Constitution, 1),
            },
        ])
    }

    fn citations(&self) -> Citations {
        let race = Citation {
            book: Book::PHB,
            page: 26,
        };
        let subrace = match self.subrace {
            HalflingSubrace::Lightfoot | HalflingSubrace::Stout => Citation {
                book: Book::PHB,
                page: 28,
            },
        };
        Citations(vec![race, subrace])
    }

    fn features(&self) -> Vec<Feature> {
        let mut features = vec![Feature {
            title: "Ability Score Increase",
            description: "Your Dexterity score increases by 2.",
            citation: Citation {
                book: Book::PHB,
                page: 28,
            },
        }];
        features.extend(match self.subrace {
            HalflingSubrace::Lightfoot => vec![Feature {
                title: "Ability Score Increase",
                description: "Your Charisma score increases by 1.",
                citation: Citation {
                    book: Book::PHB,
                    page: 28,
                },
            }],
            HalflingSubrace::Stout => vec![Feature {
                title: "Ability Score Increase",
                description: "Your Constitution score increases by 2.",
                citation: Citation {
                    book: Book::PHB,
                    page: 28,
                },
            }],
        });
        features
    }
}

impl fmt::Display for Halfling {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} Halfling", self.subrace)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_snapshot() {
        let mut rng = Pcg64::seed_from_u64(1);
        let halfling = Halfling::gen(&mut rng);
        insta::assert_yaml_snapshot!(halfling);
    }

    #[test]
    fn test_snapshot_display() {
        for subrace in HalflingSubrace::iter() {
            let halfling = Halfling { subrace };
            insta::assert_snapshot!(format!("{}", halfling));
        }
    }

    #[test]
    fn test_snapshot_abilities() {
        for subrace in HalflingSubrace::iter() {
            let halfling = Halfling { subrace };
            insta::assert_yaml_snapshot!(halfling.abilities());
        }
    }

    #[test]
    fn test_snapshot_citations() {
        for subrace in HalflingSubrace::iter() {
            let halfling = Halfling { subrace };
            insta::assert_yaml_snapshot!(halfling.citations());
        }
    }

    #[test]
    fn test_snapshot_features() {
        for subrace in HalflingSubrace::iter() {
            let halfling = Halfling { subrace };
            insta::assert_yaml_snapshot!(halfling.features());
        }
    }
}
