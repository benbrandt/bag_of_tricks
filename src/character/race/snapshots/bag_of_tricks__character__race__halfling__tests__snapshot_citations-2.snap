---
source: src/character/race/halfling.rs
expression: halfling.citations()
---
- - PHB
  - 26
- - PHB
  - 28
